drivers/docker/driver.go

```diff
@@ -0,0 +1,1250 @@
+	// configuration
+	clientConfig *base.ClientDriverConfig
+
+	// ctx is the context for the driver. It is passed to other subsystems to
+	// coordinate shutdown
+	ctx context.Context
+
+	// signalShutdown is called when the driver is shutting down and cancels the
+	// ctx passed to any subsystems
+	signalShutdown context.CancelFunc
+
+	// tasks is the in memory datastore mapping taskIDs to taskHandles
+	tasks *taskStore
+
+	// coordinator is what tracks multiple image pulls against the same docker image
+	coordinator *dockerCoordinator
+
+	// logger will log to the plugin output which is usually an 'executor.out'
+	// file located in the root of the TaskDir
+	logger hclog.Logger
+}
+
+// NewDockerDriver returns a docker implementation of a driver plugin
+func NewDockerDriver(logger hclog.Logger) drivers.DriverPlugin {
+	ctx, cancel := context.WithCancel(context.Background())
+	logger = logger.Named(pluginName)
+	return &Driver{
+		eventer:        eventer.NewEventer(ctx, logger),
+		config:         &DriverConfig{},
+		tasks:          newTaskStore(),
+		ctx:            ctx,
+		signalShutdown: cancel,
+		logger:         logger,
+	}
+}
+
+func (d *Driver) RecoverTask(handle *drivers.TaskHandle) error {
+	if _, ok := d.tasks.Get(handle.Config.ID); ok {
+		return nil
+	}
+
+	var handleState taskHandleState
+	if err := handle.GetDriverState(&handleState); err != nil {
+		return fmt.Errorf("failed to decode driver task state: %v", err)
+	}
+
+	client, _, err := d.dockerClients()
+	if err != nil {
+		return fmt.Errorf("failed to get docker client: %v", err)
+	}
+
+	container, err := client.InspectContainer(handleState.ContainerID)
+	if err != nil {
+		return fmt.Errorf("failed to inspect container for id %q: %v", handleState.ContainerID, err)
+	}
+
+	reattach, err := shared.ReattachConfigToGoPlugin(handleState.ReattachConfig)
+	if err != nil {
+		return err
+	}
+
+	dlogger, dloggerPluginClient, err := docklog.ReattachDockerLogger(reattach)
+	if err != nil {
+		return fmt.Errorf("failed to reattach to docker logger process")
+	}
+
+	h := &taskHandle{
+		client:                client,
+		waitClient:            waitClient,
+		dlogger:               dlogger,
+		dloggerPluginClient:   dloggerPluginClient,
+		logger:                d.logger.With("container_id", container.ID),
+		task:                  handle.Config,
+		containerID:           container.ID,
+		containerImage:        container.Image,
+		doneCh:                make(chan bool),
+		waitCh:                make(chan struct{}),
+		removeContainerOnExit: d.config.GC.Container,
+		net:                   handleState.DriverNetwork,
+	}
+
+	d.tasks.Set(handle.Config.ID, h)
+	go h.collectStats()
+	go h.run()
+
+	return nil
+}
+
+func (d *Driver) StartTask(cfg *drivers.TaskConfig) (*drivers.TaskHandle, *structs.DriverNetwork, error) {
+	if _, ok := d.tasks.Get(cfg.ID); ok {
+		return nil, nil, fmt.Errorf("taskConfig with ID '%s' already started", cfg.ID)
```

Comment:
`taskConfig` doesn't mean anything to end users (although this seems like it should only ever happen if there was a programming error).
```suggestion
		return nil, nil, fmt.Errorf("task with ID %q already started", cfg.ID)
```